--- lib/facets/enumerable/duplicates.rb	2008-04-20 14:28:07.000000000 +0200
+++ lib/facets/enumerable/duplicates1.rb	2008-04-20 14:47:15.000000000 +0200
@@ -16,6 +16,19 @@
     h2.keys
   end
 
+  # Returns a list of elements that occur at least n times
+  # 
+  #  [0,1,1,1,3,0,1,2,4].at_least_times( 3 ) #=> [1]
+  #  
+  #  CREDIT: Robert Dober
+  
+  def at_least_times( n = 3 )
+    h = Hash.new( 0 )
+    each {|i|
+      h[i] += 1
+    }
+    h.delete_if{|_,v| v < n}.keys
+  end
   #   #
   #   def nonuniq!
   #     raise unless respond_to?(:replace)
