--- lib/facets/enumerable/mash.rb	2008-03-08 15:10:39.000000000 +0100
+++ lib/facets/enumerable/mash1.rb	2008-04-20 14:56:29.000000000 +0200
@@ -13,13 +13,14 @@
   # NOTE: Would #correlate would be better?
 
   def mash(&yld)
-    if yld
-      inject({}) do |h, *kv| # Used to be inject({}) do |h,kv|
+    if block_given?  # faster than yld!!!
+      h = {}
+      each do |*kv| # Used to be inject({}) do |h,kv|
         r = *yld[*kv]        # The *-op works differnt from to_a on single element hash!!!
         nk, nv = *r          # Used to be nk, nv = *yld[*kv].to_a.flatten
         h[nk] = nv
-        h
       end
+      h
     else
       Enumerator.new(self,:graph)  # Used to be Hash[*self.to_a] or Hash[*self.to_a.flatten]
     end
